src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,493 @@
+{-# LANGUAGE BangPatterns    #-}
+{-# LANGUAGE DeriveFoldable  #-}
+{-# LANGUAGE DeriveFunctor   #-}
+{-# LANGUAGE MagicHash       #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns    #-}
+{- HLINT ignore "Redundant lambda" -}
+{- HLINT ignore "Use camelCase" -}
+
+-- | Incremental, in-memory run consruction
+--
+module Database.LSMTree.Internal.Run.Construction (
+    -- * Types
+    KOp
+  , RawKey
+  , RawValue (RawValue, EmptyRawValue)
+  , Operation (..)
+  , BlobRef (..)
+    -- * Pages
+    -- ** Page Builder
+  , SizedBuilder (..)
+  , pageBuilder
+    -- **  Page accumulator
+  , PageAcc
+  , paIsEmpty
+  , paIsOverfull
+  , paEmpty
+  , paAddElem
+  , paSingleton
+    -- * Exposed only for testing
+    -- ** StricterList
+  , StricterList (..)
+    -- ** BitMap
+  , BitMap (..)
+  , Bit
+  , emptyBitMap
+  , appendBit
+  , boolToBit
+    -- ** CrumbMap
+  , CrumbMap (..)
+  , Crumb
+  , emptyCrumbMap
+  , appendCrumb
+  , boolsToCrumb
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Bits
+import qualified Data.ByteString.Builder as BB
+import qualified Data.ByteString.Builder.Internal as BB
+import           Data.ByteString.Short.Internal (ShortByteString (SBS))
+import qualified Data.ByteString.Short.Internal as SBS
+import           Data.Maybe (fromMaybe, isJust)
+import           Data.Monoid (Dual (..))
+import           Data.Primitive.ByteArray
+import           Data.Word
+import qualified Database.LSMTree.Internal.Serialise as S
+import           Foreign.Ptr
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+-- | A k\/op pair with an optional blob reference.
+type KOp = (RawKey, Operation, Maybe BlobRef)
```

Comment:
For a real version we should be using `Entry` from `Database.LSMTree.Internal.Entry`. And we can probably just pass `SerialisedKey` and `Entry` as separate function args and not need a type fo the pair of the two.